core/src/main/java/com/oath/oak/EntrySet.java

```diff
@@ -0,0 +1,951 @@
+    }
+
+    /**
+     * readKey reads a key from entry at the given entry index (from off-heap).
+     * Key is returned via reusable thread-local ByteBuffer.
+     * There is no copy just a special ByteBuffer for a single key.
+     * The thread-local ByteBuffer can be reused by different threads, however as long as
+     * a thread is invoked the ByteBuffer is related solely to this thread.
+     */
+    ByteBuffer readKey(int ei) {
+        if (ei == INVALID_ENTRY_INDEX) {
+            return null;
+        }
+
+        long keyReference = getKeyReference(ei);
+        return keyRefToByteBuffer(keyReference, memoryManager);
+    }
+
+    static ByteBuffer keyRefToByteBuffer(long keyRef, MemoryManager mm) {
+        int[] keyArray = UnsafeUtils.longToInts(keyRef);
+        int blockID = getKeyBlockIDFromIntArray(keyArray);
+        int keyPosition = getPositionFromIntArray(keyArray);
+        int length = getKeyLengthFromIntArray(keyArray);
+
+        return mm.getByteBufferFromBlockID(blockID, keyPosition, length);
+    }
+
+    void setKeyRefer(int ei, OakRReference keyRef) {
+        if (ei == INVALID_ENTRY_INDEX) {
+            return;
+        }
+        long keyReference = getKeyReference(ei);
+        keyRefToOakRRef(keyReference, keyRef);
+    }
+
+    /**
+     * Sets the given external key reference (OakRReference) given the long key reference.
+     * Used while iterating, when lookUp context is not used.
+     * There is no copy just a special ByteBuffer for a single key.
+     * The thread-local ByteBuffer can be reused by different threads, however as long as
+     * a thread is invoked the ByteBuffer is related solely to this thread.
+     */
+    static void keyRefToOakRRef(long keyReference, OakRReference oakKeyRef) {
+        int[] keyArray = UnsafeUtils.longToInts(keyReference);
+        int blockID = getKeyBlockIDFromIntArray(keyArray);
+        int keyPosition = getPositionFromIntArray(keyArray);
+        int length = getKeyLengthFromIntArray(keyArray);
+
+        oakKeyRef.setReference(blockID, keyPosition, length);
+    }
+
+    /**
+     * Sets the given value reference (OakRReference) given the entry index.
+     * Returns false if: (1)there is no such entry or (2)entry has no value set
+     * Returns true otherwise
+     * a thread is invoked on the ByteBuffer is related solely to this thread.
+     */
+    boolean setValueRefer(int ei, OakRReference valueRef) {
+        if (ei == INVALID_ENTRY_INDEX) {
+            return false;
+        }
+        long valueReference = getValueReference(ei);
+        if (valueReference == INVALID_VALUE_REFERENCE) {
+            return false;
+        }
+        int[] valueArray = UnsafeUtils.longToInts(valueReference);
+        int blockID = getValueBlockIDFromIntArray(valueArray);
+        int valuePosition = getPositionFromIntArray(valueArray);
+        int length = getValueLengthFromIntArray(valueArray);
+
+        valueRef.setReference(blockID, valuePosition, length);
```

Comment:
Because keyRefToOakRRef has actually the object (OakRReference) where to put the data. And here we want to avoid creating another object just for the transmission of the output from one method to the other.